pkg/clientutil/volumes.go

```diff
@@ -0,0 +1,167 @@
+
+	// AnnDynamicallyProvisioned is added to a PV that is dynamically provisioned by kubernetes
+	// Because the annotation is defined only at k8s.io/kubernetes, copying the content instead of vendoring
+	// https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/controller/volume/persistentvolume/util/util.go#L65
+	AnnDynamicallyProvisioned = "pv.kubernetes.io/provisioned-by"
+)
+
+var VolumeResources = []string{"persistentvolumes", "persistentvolumeclaims"}
+
+func CleanupUnretainedVolumes(ctx context.Context, logger logrus.FieldLogger, c client.Client) error {
+	// We disable the PV & PVC creation so nothing creates new PV's while we delete them
+	logger.Infoln("Creating ValidatingWebhookConfiguration to disable future PV & PVC creation...")
+	if err := disablePVCreation(ctx, c); err != nil {
+		return fail.KubeClient(err, "failed to disable future PV & PVC creation.")
+	}
+
+	pvcList, pvList, err := getDynamicallyProvisionedUnretainedPvs(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	// Do not attempt to delete any pods when there are no PVs and PVCs
+	if (pvcList != nil && pvList != nil) && len(pvcList.Items) == 0 && len(pvList.Items) == 0 {
+		return nil
+	}
+
+	// Delete all Pods that use PVs. We must keep the remaining pods, otherwise
+	// we end up in a deadlock when CSI is used
+	if err := cleanupPVCUsingPods(ctx, c); err != nil {
+		return fail.KubeClient(err, "failed to clean up PV using pod from user cluster.")
+	}
+
+	// Delete PVC's
+	logger.Infoln("Deleting persistent volume claims...")
+	for _, pvc := range pvcList.Items {
+		if pvc.DeletionTimestamp == nil {
+			identifier := fmt.Sprintf("%s/%s", pvc.Namespace, pvc.Name)
+			logger.Infoln("Deleting PVC...", identifier)
+
+			if err := DeleteIfExists(ctx, c, &pvc); err != nil {
+				return fail.KubeClient(err, "failed to delete PVC from user cluster.")
+			}
+		}
+	}
+
+	return nil
+}
+
+func disablePVCreation(ctx context.Context, c client.Client) error {
+	// Prevent re-creation of PVs and PVCs by using an intentionally defunct admissionWebhook
+	creatorGetters := []reconciling.NamedValidatingWebhookConfigurationReconcilerFactory{
+		creationPreventingWebhook("", VolumeResources),
+	}
+	if err := reconciling.ReconcileValidatingWebhookConfigurations(ctx, creatorGetters, "", c); err != nil {
+		return fail.KubeClient(err, "failed to create ValidatingWebhookConfiguration to prevent creation of PVs/PVCs.")
+	}
+
+	return nil
+}
+
+func cleanupPVCUsingPods(ctx context.Context, c client.Client) error {
+	podList := &corev1.PodList{}
+	if err := c.List(ctx, podList); err != nil {
+		return fail.KubeClient(err, "failed to list Pods from user cluster.")
+	}
+
+	var pvUsingPods []*corev1.Pod
+	for idx := range podList.Items {
+		pod := &podList.Items[idx]
+		if podUsesPV(pod) {
+			pvUsingPods = append(pvUsingPods, pod)
+		}
+	}
+
+	for _, pod := range pvUsingPods {
+		if pod.DeletionTimestamp == nil {
+			if err := DeleteIfExists(ctx, c, pod); err != nil {
```

Comment:
Please use pod eviction. For more info see:
* https://github.com/kubermatic/kubeone/blob/fb64982d208763473dc9a29a62509c63f85bf8d8/pkg/nodeutils/drain.go#L78
* https://pkg.go.dev/k8s.io/kubectl/pkg/drain#Helper.EvictPod